components/calendar/src/chinese.rs

```diff
@@ -0,0 +1,1044 @@
+/// Each month starts on the date of the new moon as observed from China, meaning that months last 29
+/// or 30 days.
+///
+/// One year in the Chinese calendar is typically 12 lunar months; however, because 12 lunar months does
+/// not line up to one solar year, the Chinese calendar will add an intercalary leap month approximately
+/// every three years to keep Chinese calendar months in line with the solar year.
+///
+/// Leap months can happen after any month; the month in which a leap month occurs is based on the alignment
+/// of months with 24 solar terms into which the solar year is divided.
+///
+/// # Year and Era codes
+///
+/// Unlike the Gregorian calendar, the Chinese calendar does not traditionally count years in an infinitely
+/// increasing sequence. Instead, 10 "celestial stems" and 12 "terrestrial branches" are combined to form a
+/// cycle of year names which repeats every 60 years. However, for the purposes of calendar calculations and
+/// conversions, this module counts Chinese years in an infinite system similar to ISO, with year 1 in the
+/// calendar corresponding to the inception of the calendar, marked as 2637 BCE (ISO: -2636), and negative
+/// years marking Chinese years before February 15, 2637 BCE.
+///
+/// Because the Chinese calendar does not traditionally count years, era codes are not used in this calendar;
+/// internally, a single era code "era" is used, but this is never checked, so era code input can be anything.
+/// This Chinese calendar implementation does support a related ISO year, which marks the ISO year in which a
+/// Chinese year begins, and a cyclic year corresponding to the year in the 60 year cycle as described above.
+///
+/// For more information, suggested reading materials include:
+/// * _Calendrical Calculations_ by Reingold & Dershowitz
+/// * _The Mathematics of the Chinese Calendar_ by Helmer Aslaksen (https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.139.9311&rep=rep1&type=pdf)
+/// * Wikipedia: https://en.wikipedia.org/wiki/Chinese_calendar
+///
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Chinese;
+
+/// The inner date type used for representing [`Date`]s of [`Chinese`]. See [`Date`] and [`Chinese`] for more details.
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
+pub struct ChineseDateInner(ArithmeticDate<Chinese>);
+
+impl CalendarArithmetic for Chinese {
+    fn month_days(year: i32, month: u8) -> u8 {
+        let mid_year = Self::fixed_mid_year_from_year(year);
+        let new_year = Chinese::chinese_new_year_on_or_before_fixed_date(mid_year);
+        let approx = new_year + ((month - 1) as i64 * 29);
+        let prev_new_moon = Chinese::chinese_new_moon_before((approx + 15).as_moment());
+        let next_new_moon = Chinese::chinese_new_moon_on_or_after((approx + 15).as_moment());
+        (next_new_moon - prev_new_moon) as u8
```

Comment:
nit: probably should debug_assert that they fall in an appropriate range. (_is_ there an appropriate range?)